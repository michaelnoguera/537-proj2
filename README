CS 537 Programming Assignment 2 (Fall 2020)
Michael Noguera (noguera@cs.wisc.edu) and Julien de Castelnau (de-castelnau@cs.wisc.edu)
due 10/13/2020 at 5pm

== BUILDING ==

To build, run "make" at the root of this directory. If you specify the DEBUG environment
variable as "true" or use "make DEBUG=true", debug symbols will be included in the 
executable as well. You can also run the scan-build target to build with the Clang static
analyzer, and scan-view to view the generated output in a web browser.

Use "make clean" to get rid of object files and executables.

== USAGE ==

prodcomm takes input from standard input, so text can either be typed in and then terminated
with Ctrl-D (EOF), or it can be piped in from a file like so:

./prodcomm < test.txt

The output will be printed to standard output or it can be redirected like the input.

== PROJECT STRUCTURE ==

The functionality of prodcomm is divided into four logical modules, which serve the
 following tasks:

1. main: Initialize the three queues and call pthread_create for each worker.
2. workers: Code for Reader, Munch1, Munch2, and Writer processes, as well the helper 
            routines they require to operate (such as reader_readline)
3. queue: Queue backend and the functions that operate on queues 
          (Initialize, Enqueue, Dequeue, PrintStats).
4. stat: Statistics backend and interface, used by Queue for tracking 
         enqueue/dequeue count and time.

== TESTING FRAMEWORK ==

"make test" runs our test script which compares the output of prodcomm against a mock script
that performs the same text transformation, written in bash. Each text file in the tests 
will be tested against the expected output, and a summary of which tests passed and failed
will be printed to standard output. The stderr output from prodcomm containing statistics and
other warnings is also redirected to test.log.