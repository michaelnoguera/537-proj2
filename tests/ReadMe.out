IN*THIS*DIRECTORY,*WE*HAVE*INCLUDED*SOME*EXAMPLE*INPUT*FILES*FOR*YOU*TO*USE*IN*TESTING*PROJ2.

NOTE*THAT*YOU*CAN*REDIRECT*STANDARD*INPUT*WITH*THE*<*REDIRECT;*FOR*EXAMPLE,*TO*REDIRECT*TO*ONE_MILLION_CHAR_LINE.TXT,*YOU*CAN*ENTER

PRODCOMM*<*ONE_MILLION_CHAR_LINE.TXT

YOU*CAN*FIND*THE*FOLLOWING*FLES*IN*THIS*DIRECTORY:

TEST1*-*A*BASIC*TESTFILE
TEST2*-*AN*EMPTY*FILE
TEST3.TXT*-*A*LARGE*FILE*WITH*LARGE*INPUT.
TEN_THOUSAND_CHAR_LINE.TXT*-*A*FILE*CONTAINING*A*10,000*CHAR*LINE
ONE_HUNDRED_THOUSAND_CHAR_LINE.TXT*-*A*FILE*CONTAINING*A*100,000*CHAR*LINE
ONE_MILLION_CHAR_LINE.TXT*-*A*FILE*CONTAINING*A*1,000,000*CHAR*LINE
NO_NEWLINE.TXT*-*A*FILE*THAT*HAS*A*STRING,*THEN*TERMINATES*WITHOUT*A*NEWLINE*CHARACTER.
